cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,670 @@
+}
+
+std::tuple<rmm::device_uvector<NodeIndexT>, rmm::device_uvector<size_type>>
+records_orient_tree_traversal(device_span<SymbolT const> d_input,
+                              tree_meta_t& d_tree,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  /**
+  This algorithm assigns a unique column id to each node in the tree.
+  The row offset is the row index of the node in that column id.
+  Algorithm:
+  1. Convert node_category+fieldname to node_type.
+    a. Create a hashmap to hash field name and assign unique node id as values.
+    b. Convert the node categories to node types.
+       Node type is defined as node category enum value if it is not a field node,
+       otherwise it is the unique node id assigned by the hashmap (value shifted by #NUM_CATEGORY).
+  2. Preprocessing: Translate parent node ids after sorting by level.
+    a. sort by level
+    b. get gather map of sorted indices
+    c. translate parent_node_ids to new sorted indices
+  3. Find level boundaries.
+     copy_if index of first unique values of sorted levels.
+  4. Per-Level Processing: Propagate parent node ids for each level.
+    For each level,
+      a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+      b. stable sort by {parent_col_id, node_type}
+      c. scan sum of unique {parent_col_id, node_type}
+      d. scatter the col_id back to stable node_level order (using scatter_indices)
+    Restore original node_id order
+  5. Generate row_offset.
+    a. stable_sort by parent_col_id.
+    b. scan_by_key {parent_col_id} (required only on nodes who's parent is list)
+    c. propagate to non-list leaves from parent list node by recursion
+  **/
+  // GPU version
+  // 1. Convert node_category+fieldname to node_type.
+  using hash_table_allocator_type = rmm::mr::stream_allocator_adaptor<default_allocator<char>>;
+  using hash_map_type =
+    cuco::static_map<size_type, size_type, cuda::thread_scope_device, hash_table_allocator_type>;
+
+  constexpr size_type empty_node_index_sentinel = std::numeric_limits<size_type>::max();
+  auto num_nodes                                = d_tree.node_categories.size();
+  hash_map_type key_map{compute_hash_table_size(num_nodes),  // TODO reduce oversubscription
+                        cuco::sentinel::empty_key{empty_node_index_sentinel},
+                        cuco::sentinel::empty_value{empty_node_index_sentinel},
+                        hash_table_allocator_type{default_allocator<char>{}, stream},
+                        stream.value()};
+  auto d_hasher = [d_input          = d_input.data(),
+                   node_range_begin = d_tree.node_range_begin.data(),
+                   node_range_end   = d_tree.node_range_end.data()] __device__(auto node_id) {
+    // TODO if node_category is hashed, then no need for transform later. for field only, use string
+    // hash too.
```

Comment:
Is this still relevant? It seems we don't run into the case given we're using `insert_if` and that `get_hash_value` is only getting invoked for field name nodes?

Reply:
Yes. These TODO's for hashing are not relevant. 
I am thinking of new two-level hashing, one for field names and another for {node-level, node_category} + field hash for the entire path. This might end up faster, still need to be verified. It's possible to remove per-level processing. will remove these TODOs and add this idea as TODO in this code.

Reply:
Gotcha, so that's for a larger overhaul of the algorithm